New recipe: space-theming

p3r7 opened the pull request:
### Brief summary of what the package does

Straightforward port of the [Spacemacs theming layer](https://www.spacemacs.org/layers/+themes/theming/README.html) ([code](https://github.com/syl20bnr/spacemacs/tree/master/layers/%2Bthemes/theming)) to vanilla Emacs.

Allows overriding theme faces with the same syntax / utils as what Spacemacs provides.

For more details, see the package README and/or this [blog post](https://www.eigenbahn.com/2020/04/06/emacs-theme-override).

### Direct link to the package repository

https://github.com/p3r7/space-theming

### Your association with the package

Maintainer.

### Checklist

Please confirm with `x`:

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] `M-x checkdoc` is happy with my docstrings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

p3r7 commented:
Please note that I have a small doubt regarding the license.

Spacemacs (and all of its layers) is GPLv3 and I have released this fork under the expat license (aka MIT).

I'm no expert but there might be some compatibility issues.

riscy commented:
You're right about the compatibility issues -- the letter of the GPL requires derivative works to follow equivalence license terms, among which is that the entire source code has to be available to end users (MIT doesn't have this).  It's good that you caught that though.

On a first glance over the code, I notice some continuation characters in your comments such as here:
```elisp
(defvar space-theming-modifications '()
  "An alist of theme modifications.  Each element should \
be on the form (THEME . SPEC), where THEME is a symbol \
representing a theme, and SPEC is an alist mapping faces \
to face specs (see `defface').")
```
which shouldn't be necessary, since you can remove those and use <kbd>M-x fill-paragraph</kbd> to get the correct linebreaks both for your code and for display in the Emacs online help.

Minor:
- It may be worth it to bring in more of the [contents of the theming layer page](https://www.spacemacs.org/layers/+themes/theming/README.html) into your README (still with attribution of course) or write up a bit more about the usage in your README, just in case that page were to disappear.  Which I doubt, but you never know. :)
- Easilly -> Easily

p3r7 commented:
Tanks for the prompt feedback.

Oh, I didn't know about `fill-paragraph`, this will be quite handy (I guess I was spoiled by visual-line-mode). I've fixed that.

I've updated the license to GPLv3.

I've also taken as much as possible from the original layer documentation but provided (I hope) better insight of what faces and attributes and how to look at existing values.

riscy commented:
It's a non-blocker, but I find myself frequently recommending an SPDX license identifier (such as in your case `;; SPDX-License-Identifier: GPL-3.0-or-later`) rather than the `License: GPLv3` you have, just since it's more easily automatically detectable in the file.

I had another quick check here --


checkdoc (using version 0.6.1):
```
Warning (emacs): 
space-theming.el:40: Second line should not have indentation
Warning (emacs): 
space-theming.el:40: First line is not a complete sentence
Warning (emacs): 
space-theming.el:46: Second line should not have indentation
Warning (emacs): 
space-theming.el:46: First line is not a complete sentence
Warning (emacs): 
space-theming.el:50: Second line should not have indentation
Warning (emacs): 
space-theming.el:50: First line is not a complete sentence
Warning (emacs): 
space-theming.el:54: Second line should not have indentation
Warning (emacs): 
space-theming.el:54: First line is not a complete sentence
Warning (emacs): 
space-theming.el:105: First line is not a complete sentence
```
I think these particular checkdocs are a bit important because they determine how well your documentation will appear within various Emacs help facilities.  But I think there are some quick formatting tweaks that will fix all these:

```
(defvar space-theming-modifications '()
  "An alist of theme modifications.
Each element should be on the form (THEME . SPEC), where THEME is
a symbol representing a theme, and SPEC is an alist mapping faces
to face specs (see `defface').")

(defvar space-theming-headings-inherit-from-default '()
  "A list of themes.
All headings should inherit from the default face, or the symbol
`all'.")

(defvar space-theming-headings-same-size '()
  "A list of themes.
All headings should have the same size, or the symbol `all'.")

(defvar space-theming-headings-bold '()
  "A list of themes.
All headings should be bold, or the symbol `all'.")
```
(fill-paragraph still works with this formatting, it's pretty nice)

and...
```
(defun space-theming--theming (theme &optional _no-confirm)
  "Remove existing theming and apply customizations to THEME."
```
and I think that rewording/linebreaking fixes everything.

Ping when ready and this is a quick merge!

p3r7 commented:
Done.

I didn't know about those SPDX license identifier. That's a pretty nice system.
Do you think I could use it in my MIT/expat licensed package instead of copy-pasting the whole legal text in the header?

riscy commented:
>  Do you think I could use it in my MIT/expat licensed package instead of copy-pasting the whole legal text in the header?

Most definitely, it's in [the list](https://spdx.org/licenses/). :)
